File: app/src/main/java/com/litroenade/yunjiweather/data/repository/WeatherRepositoryFactory.java
package com.litroenade.yunjiweather.data.repository;

import com.google.gson.Gson;
import com.litroenade.yunjiweather.data.api.WeatherApiService;
import com.litroenade.yunjiweather.data.api.WeatherGatewayFactory;
import com.litroenade.yunjiweather.data.local.AppDatabase;
import com.litroenade.yunjiweather.data.local.RoomWeatherCacheGateway;

public final class WeatherRepositoryFactory {

    private WeatherRepositoryFactory() {
    }

    public static WeatherRepository createHomeRepository(
            AppDatabase database,
            WeatherApiService apiService
    ) {
        return new WeatherRepository(
                WeatherGatewayFactory.createHomeRemoteGateway(apiService),
                new RoomWeatherCacheGateway(database.weatherCacheDao(), new Gson()),
                System::currentTimeMillis
        );
    }
}

File: app/src/main/java/com/litroenade/yunjiweather/data/repository/WeatherRepositoryFactory.kt
package com.litroenade.yunjiweather.data.repository

import com.google.gson.Gson
import com.litroenade.yunjiweather.data.api.WeatherApiService
import com.litroenade.yunjiweather.data.api.WeatherGatewayFactory
import com.litroenade.yunjiweather.data.local.AppDatabase
import com.litroenade.yunjiweather.data.local.RoomWeatherCacheGateway

object WeatherRepositoryFactory {

    @JvmStatic
    fun createHomeRepository(
        database: AppDatabase,
        apiService: WeatherApiService?
    ): WeatherRepository {
        return WeatherRepository(
            WeatherGatewayFactory.createHomeRemoteGateway(apiService),
            RoomWeatherCacheGateway(database.weatherCacheDao(), Gson()),
            System::currentTimeMillis
        )
    }
}
